src/ast/passes/c_macro_expansion.cpp

```diff
@@ -0,0 +1,76 @@
+#include "ast/passes/c_macro_expansion.h"
+#include "ast/ast.h"
+#include "ast/context.h"
+#include "ast/visitor.h"
+#include "clang_parser.h"
+#include "driver.h"
+
+namespace bpftrace::ast {
+
+class CMacroExpander : public Visitor<CMacroExpander> {
+public:
+  CMacroExpander(ASTContext &ast, CDefinitions &cdefinitions)
+      : ast_(ast), cdefinitions_(cdefinitions) {};
+
+  using Visitor<CMacroExpander>::visit;
+  void visit(Expression &expr);
+
+private:
+  ASTContext &ast_;
+  CDefinitions &cdefinitions_;
+  std::unordered_set<std::string> active_;
+};
+
+void CMacroExpander::visit(Expression &expr)
+{
+  // N.B. We only support raw identifier macros. The way expansion works is
+  // that we see if an expression is a bare identifier, then attempt expansion
+  // recurisvely.
+  if (auto ident = expr.as<Identifier>()) {
+    if (cdefinitions_.macros.contains(ident->ident)) {
+      auto &value = cdefinitions_.macros[ident->ident];
+
+      // Check for recursion.
+      if (active_.contains(ident->ident)) {
+        ident->addError() << "Macro recursion limit reached: " << ident->ident
+                          << ", " << value;
+        return;
+      }
+
+      // Parse just the macro as an expression.
+      ASTContext macro(ident->ident, value);
+      Driver driver(macro);
+      auto expanded = driver.parse_expr();
+      if (!expanded) {
+        ident->addError() << "unable to expand macro as an expression: "
+                          << value;
+        return;
+      }
+
+      // Expand the macro expression in place.
+      expr.value = clone(ast_, expanded->value, ident->loc);
+
+      // Recursively visit the potentially expanded expression, ensuring that
+      // we can catch recursive expansion, per above.
+      active_.emplace(ident->ident);
```

Comment:
if you use a vector here like I did with the bpftrace macros then you can print the "path" to the recursion in the error.

Reply:
Done